x/paychan/model.go

```diff
@@ -0,0 +1,91 @@
+package paychan
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/orm"
+)
+
+var _ orm.CloneableData = (*PaymentChannel)(nil)
+
+// Validate ensures the payment channel is valid.
+func (pc *PaymentChannel) Validate() error {
+	if pc.Sender == nil {
+		return ErrMissingSender()
+	}
+	if pc.SenderPublicKey == nil {
+		return ErrMissingSenderPublicKey()
+	}
+	if pc.Recipient == nil {
+		return ErrMissingRecipient()
+	}
+	if pc.Timeout <= 0 {
+		return ErrInvalidTimeout(pc.Timeout)
+	}
+	if pc.Total == nil || !pc.Total.IsPositive() {
+		return ErrInvalidTotal(pc.Total)
+	}
+	if len(pc.Memo) > 128 {
+		return ErrInvalidMemo(pc.Memo)
+	}
+
+	// Transfer value must not be greater than the Total value represented
+	// by the PaymentChannel.
+	if pc.Transferred == nil || !pc.Transferred.IsPositive() || pc.Transferred.Compare(*pc.Total) > 0 {
+		return ErrInvalidTransferred(pc.Transferred)
+	}
+	return nil
+}
+
+// Copy returns a shallow copy of this PaymentChannel.
+func (pc PaymentChannel) Copy() orm.CloneableData {
+	return &pc
```

Comment:
req: deep clone ?

Reply:
I have implemented it like in [`escrow.Escrow.Clone`](https://github.com/iov-one/weave/blob/master/x/escrow/model.go#L49), where only a shallow copy is made. Neither addresses nor coins are cloned.

@ethanfrey is `Copy` expected to be a deep copy?

Reply:
It is supposed to not mutate the original in normal functions.

People are not "supposed to" do something like `addr[0] = 0x33`, so just copying a reference to the slice is fine.

With Coins, I believe all operations leave the original Coins untouched and return a new one. Sure, you can modify it, but you can also modify the original in another way. If we expect all datastructures to be immutible in all expected uses, I am fine with shallow clone.

If there is a good counter-argument, we can make them all deep clone.

Reply:
My understanding of deep clone is that such operation should copy the data represented by slices as well. Result should not share any memory with the original structure.

Should I change this to:
1. Deep copy (including slices and Coin),
2. Copy only top attributes values, leaving slices and Coin references pointing to original data (This is current version, so no change),
3. Copy all values, except slices. Slices can share reference between instances.

Reply:
Agree, let's stay with shallow copy and be sensitive when using it in the framework.